namerd/iface/destination/src/main/scala/io/buoyant/namerd/iface/destination/DestinationService.scala

```diff
@@ -0,0 +1,190 @@
+    override def diff[T](
+      left: Set[T],
+      right: Set[T]
+    ): Diff[Set, T] =
+      (left, right) match {
+        case (FAddress(e), _) => AddressDiff(Set.empty, Set.empty, Some(e))
+        case (_, FAddress(e)) => AddressDiff(Set.empty, Set.empty, Some(e))
+        case _ => AddressDiff(right -- left, left -- right)
+      }
+
+    override def empty[T]: Set[T] = Set.empty
+  }
+
+  case class AddressDiff[T](add: Set[T], remove: Set[T], exc: Option[Throwable] = None) extends Diff[Set, T] {
+    override def patch(coll: Set[T]): Set[T] = coll ++ add -- remove
+
+    override def map[U](f: T => U): Diff[Set, U] = AddressDiff(add.map(f), remove.map(f))
+  }
+
+  implicit val _addressDiffable = new AddressDiffable
+
+  private[this] val log = Logger.get(getClass.getName)
+  override def get(req: GetDestination): Stream[Update] = {
+    log.info(s"stream initiated from request: $req")
+    if (!req.scheme.contains("k8s")) {
+      val res = Stream.mk[Update]
+      res.close(GrpcStatus.Unimplemented("Unknown scheme: " + req.scheme.getOrElse("unknown")))
+      return res
+    }
+
+    val path = req.path match {
+      case None =>
+        val res = Stream.mk[Update]
+        res.close(GrpcStatus.Unimplemented("Empty request path"))
+        return res
+      case Some(p) => Buf.Utf8(p.substring(0, p.indexOf(".")))
+    }
+
+    val name = Path(Seq(Buf.Utf8(pathPfx), path): _*)
+    val events = interpreter.bind(Dtab.empty, name).map(_.eval).flatMap {
+      case None => Activity.value(Var.value(Addr.Bound()))
+      case Some(boundSet) =>
+        val set = Var.collect(boundSet.map(_.addr).toList).map(foldAddr)
+        Activity.value(set)
+    }.run.flatMap {
+      case Activity.Failed(e) => Var(Addr.Failed(e))
+      case Activity.Pending => Var(Addr.Pending)
+      case Activity.Ok(t) => t
+    }.changes.collect {
+      case Addr.Failed(e) => FAddress(e)
+      case Addr.Bound(addresses, _) => addresses
+    }
+
+    val updateEvents = eventFlatMap[Diff[Set, Address], Ev[Update]](events.diff, {
+      case AddressDiff(add, remove, exc) =>
+        val noEndpointsUpdate = exc match {
+          case Some(_) => Seq(Val(mkNoEndpointsUpdate(false)))
+          case None => Nil
+        }
+        val addUpdate = if (add.nonEmpty) Seq(Val(mkAddUpdate(add))) else Nil
+        val removeUpdate = if (remove.nonEmpty) Seq(Val(mkRemoveUpdate(remove))) else Nil
+        noEndpointsUpdate ++ addUpdate ++ removeUpdate
+    })
+    val totalUpdates = prependEventOnCondition[Ev[Update]](updateEvents, Val(mkNoEndpointsUpdate(true)),
+      {
+        case Val(Update(Some(Add(_)))) => true
+        case _ => false
+      })
+    VarEventStream(totalUpdates)
+  }
+
+  private[this] def foldAddr(addrs: List[Addr]): Addr = {
+    addrs.fold(Addr.Pending) {
+      // nonPending take priority
+      case (Addr.Pending, nonPending) => nonPending
+      case (nonPending, Addr.Pending) => nonPending
+      // Bound addresses combined
```

Comment:
I think you're missing the case where both addrs are bound and need to be combined.